### PR TITLE
Adding support for constrained open generics to DI

### PR DESCRIPTION
**Pull request by jbogard**
Backed-out pull request: https://github.com/dotnet/runtime/pull/34393
Previous pull request: https://github.com/dotnet/extensions/pull/536
Original discussion: https://github.com/aspnet/DependencyInjection/issues/471
Original pull request: https://github.com/aspnet/DependencyInjection/pull/635
First naive pull request: https://github.com/aspnet/DependencyInjection/pull/472

~~3rd~~ 4th time's the charm?

All of the DI packages now support all of the specification tests added.

### CODE REVIEW

File: src/libraries/Microsoft.Extensions.DependencyInjection/src/ServiceLookup/CallSiteFactory.cs
Comment (davidfowl) [inline src/libraries/Microsoft.Extensions.DependencyInjection/src/ServiceLookup/CallSiteFactory.cs:247]: When does this happen? We aimed to have 0 first chance exceptions in ASP.NET Core. This introduces some....

Comment (jbogard) [reply src/libraries/Microsoft.Extensions.DependencyInjection/src/ServiceLookup/CallSiteFactory.cs:247]: It does. There's unfortunately no way around this currently, but I have an API issue open to extend the `Type` API to have some kind of `Type.Can/TryMakeGenericType` method to use: https://github.com/dotnet/runtime/issues/28033 <- feel free to upvote :)

The behavior here we're trying to enforce is "if you ask for a single closed service, let the exception happen. If you ask for an enumerable, only return the ones that fit the constraint". The existing behavior throws *all* the time, this changes to let the exception get thrown *some* of the time.

Comment (davidfowl) [reply src/libraries/Microsoft.Extensions.DependencyInjection/src/ServiceLookup/CallSiteFactory.cs:247]: Will this make existing application registration behavior change at all? So I pay the cost of I haven't explicitly opted into this feature? Seems like yes but I want to confirm.

Comment (jbogard) [reply src/libraries/Microsoft.Extensions.DependencyInjection/src/ServiceLookup/CallSiteFactory.cs:247]: This doesn't affect registration behavior, it's only on the resolution side. It's hard to compare performance of existing behavior since the existing behavior is an unhandled exception. On the previous PR I did do some benchmarking to compare the constrained vs. unconstrained paths:

https://gist.github.com/jbogard/a2be5b903c44c8b2b3c2ef0d1e8c1e0c

**Comment by maryamariyan** [comment]
Regarding test failures refer to the list here: https://dev.azure.com/dnceng/public/_build/results?buildId=735898&view=ms.vss-test-web.build-test-results-tab

**Comment by jbogard** [comment]
@maryamariyan fixed!

**Comment by jbogard** [comment]
@maryamariyan Anything else needed from my end?

**Comment by maryamariyan** [comment]
> @maryamariyan Anything else needed from my end?

@jbogard the CI checks are green.

Let's wait on @davidfowl 's approval on this.

**Comment by dquist** [comment]
bump

**Comment by davidfowl** [comment]
Reviewing this now

File: src/libraries/Microsoft.Extensions.DependencyInjection/src/ServiceLookup/CallSiteFactory.cs
Comment (davidfowl) [reply src/libraries/Microsoft.Extensions.DependencyInjection/src/ServiceLookup/CallSiteFactory.cs:248]: I think we should re-throw the original exception. Today the container doesn't do think like make sure generic arity is the same. This will assume that every argument exception is a constraint mismatch when that isn't the case today.

We can also make this a bit more accurate by checking for those arity mismatches earlier but today we don't and this PR doesn't do that either. From the docs:

> The number of elements in typeArguments is not the same as the number of type parameters in the current generic type definition.

> -or-

> Any element of typeArguments does not satisfy the constraints specified for the corresponding type parameter of the current generic type.

>-or-

> typeArguments contains an element that is a pointer type (IsPointer returns true), a by-ref type (IsByRef returns true), or Void.

Comment (jbogard) [reply src/libraries/Microsoft.Extensions.DependencyInjection/src/ServiceLookup/CallSiteFactory.cs:248]: Well it would simplify the approach just to re-throw instead of trying to assume. Not a problem.

File: src/libraries/Microsoft.Extensions.DependencyInjection/src/ServiceLookup/CallSiteFactory.cs
Comment (davidfowl) [inline src/libraries/Microsoft.Extensions.DependencyInjection/src/ServiceLookup/CallSiteFactory.cs:246]: This exception happens once per service type right? Will get try to create the service with the wrong generic arguments more than once per service type?

Comment (jbogard) [reply src/libraries/Microsoft.Extensions.DependencyInjection/src/ServiceLookup/CallSiteFactory.cs:246]: Yeah, exactly once per closed service type. Once it's correctly resolved the `CallSite` (well, `IEnumerableCallSite`), that result is cached like any other created `CallSite`.

**Review by davidfowl — APPROVED** [review]
LGTM, lets improve this in the future by solving https://github.com/dotnet/runtime/issues/28033

**Pull request merged**
